Error handling and performance improvements

jmealo opened the pull request:
@nevill,

For your consideration:

This pull request improves performance by 14x for the following functions:
- `getEventClass()`
- `getEventName()`
- `convertToMysqlType()`

It fixes the following "warnings":
- Unterminated Statements (missing semi-colons)
- Trailing commas in arrays and objects

I fixed several very minor typos in comments and the `README.md`.

I added `TODO` comments above all `throw` and `console` statements that need attention before this module is "safe".

You did a very good job providing the `dump` functions as there was only one `console` that needed to be addressed and very few `throw` statements overall.

Obviously removing those statements won't accomplish the greater goal of making the module safe. I don't know enough about your function and tests to know where we can recover and what errors should be exposed (and how) to the end-user.

numtel commented on lib/code_map.js:
```diff
@@ -40,6 +40,45 @@ var EventCode = {
   PREVIOUS_GTIDS_LOG_EVENT: 0x23
 };
 
+var CodeEvent = [
```
Nice optimization, but now the `EventCode` object above is dead code, remove it!

jmealo replied:
Nice catch, removed the dead code in: 2422081

numtel commented on lib/sequence/binlog.js:
```diff
@@ -29,6 +29,7 @@ module.exports = function(options) {
 
   Binlog.prototype['OkPacket'] = function(packet) {
+    // TODO: Remove console statements to make module ready for widespread use
     console.log('Received one OkPacket ...');
```
Looks like we don't do anything with `OkPacket`s in ZongJi. See the `switch` at line 21.

Here's the definition in `node-mysql`: https://github.com/felixge/node-mysql/blob/master/lib/protocol/sequences/Sequence.js#L36

I would be very surprised if you could figure out a way to get that console message to actually appear. Maybe @nevill has an idea?

numtel replied:
Hmm.. if I turn the `debug` setting for the test suite to true, it outputs all a few `OkPacket`s but I never see this console message

nevill replied:
Haah, I think that's why it doesn't seem like a module.

numtel commented on test/settings/mysql.js:
```diff
@@ -4,11 +4,11 @@ module.exports = {
   connection: {
     host     : 'localhost',
     user     : 'root',
-    password : 'numtel',
```
This one...hmmm maybe put the comma at the beginning of the next line and add a comment to that `debug` will output all the incoming packets.

numtel commented:
As far as the `TODO`s for the errors, do you think you could have those all sent to the `error` event on the root ZongJi object? Would you be able to do this and write a test for it? That would definitely be much better than how it is now.

Thanks for your interest in the project! :+1:

numtel commented:
Also, while you're at it, `node-mysql` is up to version 2.9.0. 

https://www.npmjs.com/package/mysql

Can you update that version in the `package.js`?

jmealo commented:
@numtel: I have to figure out how I'll be doing error handling, my project can be run from the command line or included as a module. I think this is how ZongJi should work as well, however, I don't have time to refactor it right now, so I may need to just fix error handling as it fits my needs. I felt like these changes would benefit everyone.

I believe I addressed your comments. In regards to the TODO comments: @nevill seemed to be interested in some feedback which is why I put the TODOs in. I know how some people feel about TODOs, so I can remove them if that's what keeps this from getting merged.

Off the top of my head, this is how I think I will do it in my project:
* Constructors and blocking functions* will ``throw`` errors
* Async functions will report errors using callbacks or promises (connect)
* Errors that can occur multiple times should be emitted using event emitters (disconnected)
* When running from the command line, errors and warnings are output according to a verbosity setting
* When included as a module, nothing is output to console

What do you think of that? 

In my opinion, something like ZongJi should avoid blocking at all costs because it's hard to understand the impact to developers including it as a module in a server.

Thank you for your time reviewing my PR,
Jeff

numtel commented on lib/common.js:
```diff
@@ -545,6 +574,7 @@ exports.readMysqlValue = function(parser, column, columnSchema) {
     case MysqlTypes.DECIMAL: // Deprecated in MySQL > 5.0.3
     case MysqlTypes.NEWDATE: // Not used
     default:
+      // TODO: Remove throw statement to make module safe
```
This error can be occur multiple times and can be emitted from the ZongJi object in `lib/rows_events.js` with a `try...catch` around line 111.

jmealo replied:
I think that the performance penalty to wrap such a frequently called function in a ``try/catch`` could be significant. 

If there's an error parsing a column's value, how about we leave that column ``undefined?``  That allows you to differentiate between a ``null`` db value and a parsing error.

If there's an error parsing a row, halt parsing and emit a rowError with the raw format of the row and the error.

I believe taking a fail fast, exit early approach could work well.

numtel replied:
Hmm interesting about performance, hadn't thought of that. If we drop the `throw` here and check for a `instanceof Error` in `lib/row_events.js` we could have the same effect though.

jmealo replied:
A parsing error doesn't need to create an error or ``throw``... when parsing a row we can accumulate any parsing errors and then emit them all at once. This may require refactoring and a tighter coupling of row processing to column processing, however, every column is in a row, so making the piece of work a row, which is a set of columns, which is a set of values that are different data types, the structure is still sound.

numtel commented on lib/common.js:
```diff
@@ -75,6 +108,7 @@ var parseSetEnumTypeDef = function(type){
   }else if(type.substr(0,5).toLowerCase() === 'enum('){
     prefixLen = 5;
   }else{
+    // TODO: Remove throw statement to make module safe
```
This one should never happen... but it could also be emitted like the "Unsupported type..." error below.

jmealo replied:
As a developer, I'm more interested in the frequency of parsing errors by type so I can try to isolate the problem and find commonalities between the occurrences. Perhaps accumulating these errors up until a threshold where a report can be written to disk or an event emitted would make the most sense. I think in the end, that's what people will want to implement themselves. That and a feature flag to throw on parsing errors so one can easily debug issues would be probably make everyone pretty happy. If you're hacking any lower-level than that you've already read most of the source code.

numtel commented on index.js:
```diff
@@ -148,6 +151,7 @@ ZongJi.prototype._fetchTableInfo = function(tableMapEvent, next) {
     tableMapEvent.schemaName, tableMapEvent.tableName);
 
   this.ctrlConnection.query(sql, function(err, rows) {
+    // TODO: Remove throws to make the module safe
```
This one can happen multiple times (gets ran for each `TableMap` packet) and can be emitted from the `ZongJi` object.

numtel commented on index.js:
```diff
@@ -99,6 +99,7 @@ ZongJi.prototype._isChecksumEnabled = function(next) {
         // MySQL < 5.6.2 does not support @@GLOBAL.binlog_checksum
         return next(false);
       }
+      // TODO: Remove throws to make the module safe
```
This `throw` and the 2 following (@ line 115 and line 129) could be reported with a callback on the `ZongJi` constructor. This would require adding a third argument to the constructor.

jmealo replied:
Constructors are safe to call and are considered ready to go once they return a value.

The connection is a property of the object, not the object itself. I feel that the connection should not dictate the lifecycle of the object (or a process) the developer should.

If we're going to allow/encourage the developer to implement their own connection handling, it probably shouldn't be in the constructor. Sure, you could do it, but it will be deeply nested and very clunky.

numtel commented:
Ok, I've made comments for each of those `TODO`s. These should be pretty simple to implement. If you don't have time, let me know and I can come up with that patch.

As far as the verbosity setting on the output, I am probably not going to tackle that one myself.

jmealo commented:
What I got out of this was:
* The only real errors that a user cares about are connection/server related and should use event emitters.
* Parsing errors on one or more columns should result in that row being ignored and/or emitting an error. The developer should decide.

A developer can only address a parsing errors if the have access to the row as a whole, rather than just a column. If we have to parse the packet ourselves as part of handling an error, that'd be a mess. That's why I think that the parser should parse as many columns as it can before throwing the row error.

Allowing them to handle anything at the column level for error handling purposes is a missed opportunity to allow filters by column name and datatype instead for roughly the same amount of work, right?

nevill commented on test/types.js:
```diff
@@ -298,11 +298,11 @@ defineTypeTest('datetime_fraction', [
 ], [
   ['"1970-01-01 00:00:01.123"'],
   ['"2038-01-18 03:14:07.900"'],
-  ['"2014-12-27 01:07:08.001"'],
```
I'd love to have the comma, which is at the last element of an array, or an object (a hash). In practice, it's easier for you to cut & paste, or swap the line up or down (in Sublime Text, the shortcut is Command + Ctrl + :arrow_up: / :arrow_down: ) , even it's the last element. I think we kept a uniform style in this project.

jmealo replied:
@nevill I see a lot of people using the comma, some linters complain if it's there and others complain if it's not. I'll revert these eventually.

nevill commented:
@jmealo About error handling, I don't have much idea now, I need some real use cases. But at least, I'd like to have all the errors emitted by ZongJi object at the end. E.g., we can do ```zongji.on('error', errorHandler)``` in the client code.

jmealo commented:
@nevill @numtel I need the connection related errors to pass the binlog position so that upon reconnection we can resume from that position. This is important because the process might crash (unrelated to ZongJi) so keeping a runny tally of the last binlog position processed is important. Also, since we can only set the binLog position before calling ``start`` we'll have to tear down and reconnect anyway.

I want the parsing errors to be warnings and not cause the process to crash. I still think setting columns that can't be parse to undefined and optionally allowing those to go to a log somewhere perhaps with a threshold (If x number of parsing errors occur, throw an exception).

Does that sound reasonable?

numtel commented:
Putting the binlog position in the error sounds good to me. I agree with @nevill that having all the errors emitted by the ZongJi object makes the most sense.

Have you encountered instances of field parsing errors? If so, those should have issues made and be fixed along with tests to make sure that they don't happen again.

numtel commented on lib/common.js:
```diff
@@ -1,35 +1,68 @@
+  GEOMETRY: 255
+};
+
+var MysqlCodeTypes = {
```
Instead of writing this out hardcoded twice, I think it would be better to generate the second object.

```javascript
// Provide flipped object for quick look-up
var MysqlCodeTypes = {}
Object.keys(MysqlTypes).forEach(function(typeId) {
  MysqlCodeTypes[MysqlTypes[typeId]] = typeId;
});
```

jmealo replied:
V8 can do a better job of optimizing a fixed length object literal than it will doing it dynamically, but that's not a good argument to keep it this way.

The only solid reason why I generally do this is so that code introspection/autocomplete in IDEs will be able to find the values.

I have used those types of functions int he past for larger lists that were remotely loaded. I'd say we should just move all these definitions into a file where there's no code.

numtel commented:
I am interested where you obtained the figures for performance improvements. If you have written a test suite for performance, it would be helpful to have it included in the repo. There has been some interest in obtaining such details: https://github.com/nevill/zongji/issues/18

It would be really nice to have a script that can provide an output of events per second for a high load of queries performed against the MySQL server in order to gain insight into performance improvements over the entire package.

jmealo commented:
@numtel: I ran benchmarks against individual functions in Chrome Canary using jsperf. I ran them against the code present in master, versus the code in my PR and refined them until they were as performant as possible while still being readable.

For benchmarking, I use ``sysbench`` to thrash MySQL. I have found that with my improvements (possibly without), on a 8 core system, you cannot generate enough MySQL traffic using the advanced ``sysbench`` to stress out Zongjii when running sysbench on the same system.

In other words, even if you have 6 cores of MySQL running at 100%, if you leave one core to run sysbench and another core to run Zongji, Zongji will not get near 100%.

To explain this differently, and I apologize for not keeping better records, I think the performance of zongji can best be explained (when running on the same machine without bandwidth latency).

1. Restart MySQL and Zongji.
2. Run sysbench.
3. At the end of the test, record the CPU time for each.

It's good to measure latency and "time to catch up" as well.

My project uses two Zongji listeners on the same server to eliminate outliers in memory and CPU usage.

Whenever I add instrumentation to measure latency I'll try to pass it along. In the meantime, if you guys could comment some of the code or just add what each file is for that'd be helpful. I know what any piece does, but it's not clear how it fits together.

jmealo commented:
@numtel: I haven't seen any instances of the parsing errors being thrown, but that's a programmer error inside of the library. I think it needs to be written that a parsing error can be handled by dropping the column and/or row (if it's a PK).

jmealo commented:
I'm working on streaming changes from MySQL, Redis, MongoDB and PostgreSQL using binary logging or streaming replication as available in each. I'd like to use Zongji but I may have to fork it to get it to behave, that just seems like a waste because we're in agreement about what needs to get done to make it module safe, just wanted to give you an update of how I intend to use it:

https://github.com/jmealo/lapidus

Performance and reliability are my priorities.

This is just a first iteration:
https://github.com/jmealo/lapidus/blob/master/lib/mysql.js

You'll see I give the option to wrap your event callbacks so that they happen either at the end of the current event loop iteration or the beginning of the next. It's not advisable but if they wanted to block they could block.

The important part is getting the streaming part working so you can pause the stream when a new table is created so that I can figure out what the primary key is, otherwise I won't know for the first insert on each table.

numtel commented:
Sorry about not merging sooner. I figured that you were planning on writing the changes to the error handling in this pull request and that this was still a work in progress.

Very cool to hear about that performance overall.

numtel commented:
@jmealo @nevill 

I've created a branch called `error_handling` that includes the commits from this pull request and adds 3 commits for emitting errors from the ZongJi object.

https://github.com/nevill/zongji/commits/error_handling

The only operational change is that if a row event contains a column that is not handled by `Common.readMysqlValue`, the value will be `undefined` in the event and the ZongJi object will emit an error with a message like, `Error: Unsupported type "BIT" on column "col1" of the table "type_bit" in the database "zongji_test"`.

I've also included a commit bumping the package version to 0.3.3. I think this is ready to be released now.

jmealo commented:
Sounds great!
On Oct 26, 2015 11:05 AM, "Ben Green" <notifications@github.com> wrote:

> @jmealo <https://github.com/jmealo> @nevill <https://github.com/nevill>
>
> I've created a branch called error_handling that includes the commits
> from this pull request and adds 3 commits for emitting errors from the
> ZongJi object.
>
> https://github.com/nevill/zongji/commits/error_handling
>
> The only operational change is that if a row event contains a column that
> is not handled by Common.readMysqlValue, the value will be undefined in
> the event and the ZongJi object will emit an error with a message like, Error:
> Unsupported type "BIT" on column "col1" of the table "type_bit" in the
> database "zongji_test".
>
> I've also included a commit bumping the package version to 0.3.3. I think
> this is ready to be released now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nevill/zongji/pull/19#issuecomment-151167128>.
>

jmealo commented:
@numtel: I'm going to test this out. I'd like to implement health checking in https://github.com/jarvusinnovations/lapidus which will show the health of any configured backends (mysql, postgresql, mongodb).

I have been running lapidus/zongji in production and I haven't run into an issue with Lapidus crashing... however, Lapidus runs on the same server as my MySQL instance and the two are never really restarted independently of one another.

I'll have to test if restarting MySQL causes an issue and how long MySQL can be down without Zongji timing out or throwing an error.

I don't think the things necessary to implement a health check need to be added to Zongji, however, ``connecting``, ``connected``, ``disconnected``, ``retrying`` events should be relayed from the underlying connection.

jmealo commented:
@numtel is it safe to send a ping over the control connection and the usual connection?

jmealo commented:
@numtel closing/re-opening to see if Travis will run on the PR
